### PR TITLE
[turnstile] Complete tutorial: integrating Turnstile with Bot Management and WAF

### PR DESCRIPTION
**Pull request by kristianfreeman**
### Documentation checklist

- [ ] The [documentation style guide](https://developers.cloudflare.com/style-guide/) has been adhered to.
- [ ] If a larger change - such as adding a new page- an issue has been opened in relation to any incorrect or out of date information that this PR fixes.
- [ ] Files which have changed name or location have been allocated [redirects](https://developers.cloudflare.com/pages/configuration/redirects/#per-file).

### CODE REVIEW

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+  - Web Application Firewall
+  - Bot Management
+languages:
+  - JavaScript
+sidebar:
+  order: 3
+
+---
+
+Cloudflare's security tools offer powerful protection for web applications. In this tutorial, we'll explore how to integrate Cloudflare Turnstile, Web Application Firewall (WAF), and Bot Management into an existing authentication system. This combination creates a robust defense against various threats, including automated attacks and malicious login attempts.
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
This tutorial will guide you on how to integrate Cloudflare Turnstile, [Web Application Firewall (WAF)](/waf/), and [Bot Management](/bots/get-started/bm-subscription/) into an existing authentication system. This combination creates a robust defense against various threats, including automated attacks and malicious login attempts.
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+## Overview
+
+To use WAF and Bot Management, your site must have its DNS pointing through Cloudflare. However, Turnstile can be used independently on any site, even those not on Cloudflare's network. We'll cover how to implement all three, but you can focus on Turnstile if your site isn't on Cloudflare.
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
To use WAF and Bot Management, your site must have its DNS pointing through Cloudflare. However, Turnstile can be used independently on any site including those not on Cloudflare's network. This tutorial will cover how to implement all three products, but you can focus on Turnstile if your site is not on Cloudflare's network.
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+These tools work well together, by working on different layers of the application:
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
WAF, Bot Management, and Turnstile work well together by operating on different layers of the application:
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+- WAF filters malicious traffic based on network signals.
+- Bot Management analyzes requests to identify and mitigate automated threats.
+- Turnstile examines client-side and browser signals to distinguish between human users and bots.
+
+By combining server-side (WAF and Bot Management) and client-side (Turnstile) security measures, you can combine together multiple layers of defense to create a protection system that is difficult for attackers to circumvent.
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
By combining server-side (WAF and Bot Management) and client-side (Turnstile) security measures, you can combine multiple layers of defense to create a protection system that is difficult for attackers to circumvent.
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+## Prerequisites
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
## Before you begin
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+Ensure you have:
+
+- A Cloudflare account with access to WAF and Bot Management (if using)
+- An existing JavaScript/TypeScript-based route handling authentication
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
- You must have a Cloudflare account with access to WAF and Bot Management (if using).
- An existing JavaScript/TypeScript-based route handling authentication.
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+In this example, we'll use a simple login form written in plain HTML to demonstrate how to integrate Turnstile into your application. In the backend, we will use a stubbed out authentication route, written in TypeScript, to handle the login request. You may replace this with the language of your choice. As long as your language or framework is able to make an external HTTP request to Turnstile's API, you can integrate Turnstile into your application.
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
This tutorial uses a simple login form written in plain HTML to demonstrate how to integrate Turnstile into your application. In the backend, a stubbed out authentication route, written in TypeScript, will handle the login request. You may replace this with the language of your choice. As long as your language or framework is able to make an external HTTP request to [Turnstile's API](/api/operations/accounts-turnstile-widget-create), you can integrate Turnstile into your application.
```
Is this the right link to use for API?

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+## Configuring WAF and Bot Management
+
+If your site is on Cloudflare's network and under an Enterprise plan, set up WAF and Bot Management:
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
If your site is on Cloudflare's network and subscribed to an Enterprise plan, you must configure WAF and Bot Management.
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+## Prerequisites
+
+Ensure you have:
+
+- A Cloudflare account with access to WAF and Bot Management (if using)
+- An existing JavaScript/TypeScript-based route handling authentication
+
+In this example, we'll use a simple login form written in plain HTML to demonstrate how to integrate Turnstile into your application. In the backend, we will use a stubbed out authentication route, written in TypeScript, to handle the login request. You may replace this with the language of your choice. As long as your language or framework is able to make an external HTTP request to Turnstile's API, you can integrate Turnstile into your application.
+
+## Configuring WAF and Bot Management
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
## Configure WAF and Bot Management
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+### Issue challenges for potential bot traffic
+
+1. Log in to your Cloudflare dashboard.
+2. Navigate to Security > WAF.
+3. Create a new custom WAF rule, clicking "Edit expression":
+   - Field: "Bot Score"
+	 - Operator: "less than or equal to"
+	 - Value: "30"
+   - Action: "Managed Challenge"
+
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
1. Log in to the [Cloudflare dashboard](https://dash.cloudflare.com/) and select your account.
2. Go to **Security** > **WAF**.
3. Create a new custom WAF rule by selecting "Edit expression":
   - Field: "Bot Score"
	 - Operator: "less than or equal to"
	 - Value: "30"
   - Action: "Managed Challenge"

```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+![WAF rule to challenge bot traffic](~/assets/images/turnstile/complete-tut--rule-definition.png)
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]: This screenshot might not be necessary and can become out of date.
```suggestion
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+
+This configuration challenges requests with a low bot score, leveraging network signals to identify potential threats before they reach your application. You may customize the score threshold based on your specific use case.
+
+## Setting up Cloudflare Turnstile
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
## Set up Cloudflare Turnstile
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+
+Turnstile can be used on any site, regardless of whether it's on Cloudflare's network:
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
Turnstile can be used on any site, regardless of whether it is on Cloudflare's network:
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+
+1. In your Cloudflare dashboard, go to Security > Turnstile.
+2. Create a new Turnstile site.
+3. Add your domain to the Turnstile configuration.
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
1.Log in to the [Cloudflare dashboard](https://dash.cloudflare.com/) and select your account.
2. Go to **Security** > **Turnstile**.
3. Select **Add a widget** and fill out the necessary information.
4. Add your domain to the Turnstile configuration.
5. Select **Create**.
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+
+Turnstile adds an extra layer of security by analyzing browser and client-side signals, complementing the server-side checks performed by WAF and Bot Management.
+
+![Turnstile admin](~/assets/images/turnstile/complete-tut--admin.png)
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]: This is probably not necessary either.
```suggestion
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+### Enable the option to use the existing clearance cookie
+
+If your site is on Cloudflare, enable the option to use the existing clearance cookie in Turnstile settings. This integration allows Turnstile to use the Clearance Cookie as part of its determination of whether a user should receive a challenge. This integration is optional, but if you already are making use of WAF and Bot Management, it's recommended to enable this option.
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
If your site is on Cloudflare, you can enable the option to use the existing [clearance cookie](/turnstile/concepts/pre-clearance-support/) in Turnstile's settings. This integration allows Turnstile to use the clearance cookie as part of its determination of whether a user should receive a challenge. This integration is optional, but recommended if you already are using WAF and Bot Management.
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+## Integrating Turnstile into Your Application
+
+There are two components to integrating Turnstile into your application: the Turnstile widget and the server-side validation logic.
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
There are two components to implementing Turnstile into your application: the Turnstile widget and the server-side validation logic.
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+  <div class="cf-turnstile" data-sitekey="YOUR_SITE_KEY"></div>
+  <button type="submit">Log in</button>
+</form>
+
+<script src="https://challenges.cloudflare.com/turnstile/v0/api.js" async defer></script>
+```
+
+Replace `YOUR_SITE_KEY` with your actual Turnstile site key.
+
+![Turnstile widget in a login form](~/assets/images/turnstile/complete-tut--frontend.png)
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]: This is also not necessary. Might be a bit rendundant?
```suggestion
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+
+## Handling the Login Request
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
## Handle the login request
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+    });
+  }
+}
+
+async function checkCredentials(username: string, password: string): Promise<boolean> {
+  // Your existing credential checking logic
+}
+```
+
+This setup ensures that the Turnstile token is validated server-side before proceeding with the login process, adding an extra layer of security based on client-side signals.
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
This setup ensures that the Turnstile token is validated on the server-side before proceeding with the login process, adding an extra layer of security based on client-side signals.
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+
+## Testing
+
+After deployment, you will want to test your integration. Because your bot score will be low, you will probably not receive a challenge. However, you can add additional rules as needed to force a redirect to the challenge page. Some options to do this are:
+
+1) Add a WAF rule that always forwards your IP address to the challenge page.
+2) Add a WAF rule that checks for the presence of a query parameter, such as `?challenge=true`.
+
+## Best Practices
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
## Best practices
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+
+1. Always validate the Turnstile token server-side before checking credentials.
+2. Use environment variables to store sensitive information like your Turnstile secret key.
+3. Implement proper error handling and logging to monitor for potential security issues.
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
1. Always validate the Turnstile token on the server-side before checking credentials.
2. Use environment variables to store sensitive information like your Turnstile secret key.
3. Implement proper error handling and logging to monitor for potential security issues.
```

File: src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+
+By combining Turnstile with WAF and Bot Management, you can create a system that secures your application at the network layer, while also providing an extra layer of protection using client-side signals. This approach makes it significantly more difficult for malicious actors to automate attacks against your login system.
+
+## Resources
+
+If you're interested in customizing Turnstile, here are some additional resources to check out next:
```
Comment (patriciasantaana) [inline src/content/docs/turnstile/tutorials/integrating-turnstile-waf-and-bot-management.mdx]:
```suggestion
If you are interested in customizing Turnstile, refer to the resources below for more information:
```

**Comment by kristianfreeman** [comment]
There was a list rendering issue on the preview build that I just fixed. It's ready to go now! @patriciasantaana can you re-review? This is ready to merge whenever also, per PM

**Review by patriciasantaana — APPROVED** [review]

**Pull request merged**